server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.Diffable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This is a {@code Map<String, String>} that implements AbstractDiffable so it
+ * can be used for cluster state purposes
+ */
+public class DiffableStringMap extends AbstractMap<String, String> implements Diffable<DiffableStringMap> {
+
+    private final Map<String, String> innerMap;
+
+    DiffableStringMap(final Map<String, String> map) {
+        this.innerMap = map;
+    }
+
+    @SuppressWarnings("unchecked")
+    DiffableStringMap(final StreamInput in) throws IOException {
+        this.innerMap = (Map<String, String>) (Map) in.readMap();
+    }
+
+    @Override
+    public String put(String key, String value) {
+        return innerMap.put(key, value);
+    }
+
+    @Override
+    public Set<Entry<String, String>> entrySet() {
+        return innerMap.entrySet();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeMap((Map<String, Object>) (Map) innerMap);
+    }
+
+    @Override
+    public Diff<DiffableStringMap> diff(DiffableStringMap previousState) {
+        return new DiffableStringMapDiff(previousState, this);
+    }
+
+    public static Diff<DiffableStringMap> readDiffFrom(StreamInput in) throws IOException {
+        return new DiffableStringMapDiff(in);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```

Comment:
I don't think it is necessary to overwrite equals(), hashCode() and toString() here? `AbstractMap` already implements these methods.